lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -261,13 +263,50 @@ fn connect_rdp_inner(
         Ok(())
     });
 
+    let tdp_sd_info_request = Box::new(move |req: SharedDirectoryInfoRequest| -> RdpResult<()> {
+        debug!("sending: {:?}", req);
+        // Create C compatible string from req.path
+        match CString::new(req.path.clone()) {
```

Comment:
Don't you already have a C-compatible string from Go's `C.String()` call?

Reply:
> Don't you already have a C-compatible string from Go's C.String() call?

I'm not sure which call you're talking about. The call to this function originates [here](https://github.com/gravitational/teleport/blob/6bfc7a3be4816f96b3ec57bd2057eb820ef3d946/lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs#L278).

> I'm not certain the memory management is correct here.

We create a `CString` here, which gets passed to the Go function [here](https://github.com/gravitational/teleport/blob/6bfc7a3be4816f96b3ec57bd2057eb820ef3d946/lib/srv/desktop/rdp/rdpclient/client.go#L490) and copied into a Go `string`, at which point we free the memory (`let _ = CString::from_raw(c_string);`).

This is based on the example [here](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.into_raw)

Reply:
Okay, so this is what's going on:

We have an owned String
We clone it and append a null terminator when we create the CString, we still own it
We convert it into a raw pointer to pass to a Go callback
We convert the raw pointer back into a CString to ensure it's dropped and the memory is freed.
It seems to be that 3 and 4 are slightly convoluted, why do we convert into a raw pointer and then back to drop it, can't we just use as_ptr() on the CString to get a raw pointer to pass to the Go callback? We want to retain ownership to ensure it's freed anyway right? Other than that it looks correct.

Reply:
@zmb3 @ibeckermayer

Reply:
Thank you Joel, master of the FFI. 😀